x/wasm/client/cli/tx.go

```diff
@@ -415,18 +423,61 @@ func GrantAuthorizationCmd() *cobra.Command {
 			if exp == 0 {
 				return errors.New("expiration must be set")
 			}
-			_ = clientCtx
-			_ = grantee
-			_ = msgs
-			_ = once
-			_ = contract
 
-			return errors.New("not implemented")
+			var limit types.ContractAuthzLimitX
+			switch {
+			case maxFundsStr != "" && maxCalls != 0:
+				maxFunds, err := sdk.ParseCoinsNormalized(maxFundsStr)
+				if err != nil {
+					return fmt.Errorf("max funds: %s", err)
+				}
+				limit = types.NewCombinedLimit(maxCalls, maxFunds)
+			case maxFundsStr != "":
+				maxFunds, err := sdk.ParseCoinsNormalized(maxFundsStr)
+				if err != nil {
+					return fmt.Errorf("max funds: %s", err)
+				}
+				limit = types.NewMaxFundsLimit(maxFunds)
+			case maxCalls != 0:
+				limit = types.NewMaxCallsLimit(maxCalls)
+			default:
+				limit = types.UndefinedLimit{}
+			}
+
+			var filter types.ContractAuthzFilterX
+			switch len(msgs) {
+			case 0:
+				filter = types.NewAllowAllMessagesFilter()
+			default:
+				filter = types.NewAcceptedMessageKeysFilter(msgs...)
+			}
+
+			grant, err := types.NewContractGrant(contract, limit, filter)
+			if err != nil {
+				return err
+			}
+
+			var authorization authz.Authorization
+			switch args[1] {
+			case "execution":
+				authorization = types.NewContractExecutionAuthorization(*grant)
+			case "migration":
+				authorization = types.NewContractMigrationAuthorization(*grant)
+			default:
+				return fmt.Errorf("%s authorization type not supported", args[1])
+			}
+
+			grantMsg, err := authz.NewMsgGrant(clientCtx.GetFromAddress(), grantee, authorization, time.Unix(0, exp))
+			if err != nil {
+				return err
+			}
+			return tx.GenerateOrBroadcastTxCLI(clientCtx, cmd.Flags(), grantMsg)
 		},
 	}
 	flags.AddTxFlagsToCmd(cmd)
 	cmd.Flags().StringSlice(flagAllowedMsgs, []string{}, "Allowed msgs")
-	cmd.Flags().Bool(flagRunOnce, false, "Allow to execute only once")
+	cmd.Flags().Uint64(flagMaxCalls, 0, "Maximal number of calls to the contract")
```

Comment:
`--allow-msgs` is not the best name anymore. Not sure if `--allow-contract-msg-keys` is better but we need to differentiate between the raw json message (byte equal) and the json object keys. CLI doc is important here.
For example:
With a contract payload message: `{"foo":"bar"}`
the key option would be `foo` 
and the raw msg option `{"foo":"bar"}`